extensions/amp-analytics/0.1/instrumentation.js

```diff
@@ -148,6 +187,71 @@ class InstrumentationService {
   }
 
   /**
+   * Register for a listener to be called when the boundaries specified in
+   * config are reached.
+   * @param {!JSONObject} config the config that specifies the boundaries.
+   * @param {Function} listener
+   * @private
+   */
+  registerScrollTrigger_(config, listener) {
+    if (!Array.isArray(config['vertical-boundaries']) &&
+        !Array.isArray(config['horizontal-boundaries'])) {
+      return;
+    }
+
+    // Ensure that the scroll events are being listened to.
+    if (!this.scrollHandlerRegistered_) {
+      this.scrollHandlerRegistered_ = true;
+      this.viewport_.onChanged(this.onScroll_.bind(this));
+    }
+
+    const triggerScrollEvents = function(bounds, scrollPos) {
+      if (!scrollPos) {
+        return;
+      }
+      for (const b in bounds) {
+        if (b > scrollPos || bounds[b]) {
+          continue;
+        }
+        bounds[b] = true;
+        listener(new AnalyticsEvent(AnalyticsEventType.SCROLL));
+      }
+    };
+
+    const boundsV = this.normalizeBoundaries_(config['vertical-boundaries']);
+    const boundsH = this.normalizeBoundaries_(config['horizontal-boundaries']);
+    this.scrollObservable_.add(e => {
+      triggerScrollEvents(boundsV,
+          (e.top + e.height) * 100 / this.viewport_.getScrollHeight());
+      triggerScrollEvents(boundsH,
+          (e.left + e.width) * 100 / this.viewport_.getScrollWidth());
+    });
+  }
+
+  /**
```

Comment:
Added comments here and in other places. Please let me know if you would like to see more.